docs/tutorial.rst

```diff
@@ -1,4 +1,209 @@
 Tutorial
 ========
 
-*Coming soon!*
+In this document, we are going to develop an application that will
+convert a given time (and date) between two time zones.
+The user will enter a date and a time,
+and select two time zones from a list of all available time zones,
+one for the source location and one for the destination location.
+
+.. note::
+
+   Time zones are represented in "region/location" format,
+   as in "Australia/Tasmania".
+   While most such codes have two components,
+   some contain only one (like "UTC"),
+   and some contain more than two (like "America/North_Dakota/New_Salem".
+   Also note that spaces in region and location names are replaced
+   with underscores.
+   See the `Wikipedia list`_ for more information.
+
+
+Prerequisites
+-------------
+
+First of all, we have to install the dependencies.
+These are: *clastic* (obviously) and *dateutil*.
+Note that the PyPI package name for *dateutil* is *python-dateutil*::
+
+  pip install clastic python-dateutil
+
+
+
+Getting started
+---------------
+
+Here is the initial version of our application that only displays
+the form but doesn't handle the submitted data:
+
+.. code-block:: python
+
+   import os
+
+   from clastic import Application
+   from clastic.render import AshesRenderFactory
+   from dateutil import zoneinfo
+
+
+   def home():
+       zone_info = zoneinfo.get_zonefile_instance()
+       zone_names = zone_info.zones.keys()
+       entries = ((zone.split("/")[-1], zone) for zone in zone_names)
+       zones = [
+           {"location": location.replace("_", " "), "zone": zone}
+           for location, zone in sorted(entries)
+       ]
+       return {"zones": zones}
+
+
+   def create_app():
+       routes = [("/", home, "home.html")]
+       cur_path = os.path.dirname(__file__)
+       render_factory = AshesRenderFactory(cur_path)
+       return Application(routes, render_factory=render_factory)
+
+
+   if __name__ == "__main__":
+       app = create_app()
```

Comment:
I recommend moving this line up, outside of the `__name__ == "__main__"` block. `create_app` is idempotent and stateless. And this way, it's easily importable for other WSGI servers, so I think it's actually best practice.

Reply:
I was aware of the WSGI issue and thought about doing it the way you've described. But won't that hurt testing? I'd like to look into this some more but I guess you're in a much better position than me to evaluate it.

On a related note, how about arranging the example to use WSGI (for example wsgiref) instead of `.serve()`?

Reply:
I'm not sure how it would impact testing. Tests can instantiate additional instances of this application without impacting each other. For stateful app creation, you're right; I don't instantiate the app at module level, that's done in a separate file just for the wsgi server ([example](https://github.com/hatnote/montage/blob/master/app.py)).

As far as wsgiref, I do recommend `.serve()` instead. It sets up a lot of goodies that make it pretty useful, like the [werkzeug debug middleware](https://werkzeug.palletsprojects.com/en/1.0.x/debug/) and clastic meta app.

Reply:
OK, got it about wsgiref. How do we proceed about app creation? Do you prefer moving it to the module level, or leaving it to another module for wsgi? In the tutorial, I mean.

Reply:
Never mind, I've moved it to module level and rearranged the narration in the tutorial. I think it's OK now.